### PR TITLE
downloading and resizing is done

### PR DESCRIPTION
**Pull request by excellencetechnologies**

### CODE REVIEW

File: service/request.js [hunk 1]
```diff
@@ -23,3 +23,30 @@ exports.request = function (body, headers, url, callback) {
         }
     });
 }
+
+exports.resize = function (url, APP_ID, callback) {
+    mystring = APP_ID.replace('.', '');
```
Comment (upexcel) [inline service/request.js]: can you write code to replace all special characters

Comment (upexcel) [inline service/request.js]: use proper variables names below everywhere

File: routes/category.js [hunk 2]
```diff
@@ -12,64 +10,74 @@ var redis = require("redis"),
...
+                var url = URL + '/category/products/';
+                request_.request(body, headers, url, function (req, response, msg) {
+                    if (msg === ERROR) {
+                        res.json({status: 0, statuscode: ERR_STATUS, error: response});
+                    } else if (req.statusCode === ERR_STATUS) {
+                        res.json({status: 0, statuscode: req.statusCode, body: response});
+                    } else {
+                        var resp = JSON.parse(response);
+                        var image_u = resp.data;
+                        if (image_u !== undefined) {
```
Comment (excellencetechnologies) [inline routes/category.js]: write else condition also for this

File: routes/category.js [hunk 3]
```diff
@@ -12,64 +10,74 @@ var redis = require("redis"),
...
+                var headers = {APP_ID: APP_ID};
+                var url = URL + '/category/products/';
+                request_.request(body, headers, url, function (req, response, msg) {
+                    if (msg === ERROR) {
+                        res.json({status: 0, statuscode: ERR_STATUS, error: response});
+                    } else if (req.statusCode === ERR_STATUS) {
+                        res.json({status: 0, statuscode: req.statusCode, body: response});
+                    } else {
+                        var resp = JSON.parse(response);
+                        var image_u = resp.data;
```
Comment (excellencetechnologies) [inline routes/category.js]: rename image_u ==> categoryData

File: routes/category.js [hunk 4]
```diff
@@ -12,64 +10,74 @@ var redis = require("redis"),
...
-                                "page": page,
-                                "limit": limit,
-                                "body": response,
-                                "status": 1,
-                                "statuscode": req.statusCode
+                            res.json({status: 0, statuscode: '500', body: ERROR});
+                        }
+
+                        function processData(item, key, callback) {
+                            var image_url = item.data.media_images[0];
```
Comment (excellencetechnologies) [inline routes/category.js]: i think api has change confirm with vishal

File: routes/category.js [hunk 5]
```diff
@@ -12,64 +10,74 @@ var redis = require("redis"),
...
+                            request_.resize(image_url, APP_ID, function (status, response_, image_name) {
+                                image_url = image_name;
```
Comment (excellencetechnologies) [inline routes/category.js]: remove this line directly assign in item.data

File: routes/home.js [hunk 6]
```diff
@@ -30,14 +28,22 @@ router.post('/products', function (req, res) {
...
-                            'type': type,
-                            "data": response,
-                            "status": 1,
-                            "statuscode": req.statusCode
-                        });
-                        client.expire('products_' + type, config.PRODUCT_EXPIRESAT);
+                        var resp = JSON.parse(response);
+                        for (i = 0; i < resp.data[0].data.media_images.length; i++) {
+                            var url = resp.data[0].data.media_images[i];
+                            request_.resize(url, APP_ID, function (status, response_, image_name) {
```
Comment (excellencetechnologies) [inline routes/home.js]: same problem here we need to use async

File: routes/home.js [hunk 7]
```diff
@@ -51,72 +57,58 @@ router.post('/categories', function (req, res) {
...
-    } else {
-        res.json({status: 0, statuscode: ERR_STATUS, msg: INVALID});
-    }
+    client.hgetall('slider', function (err, object) {
+        if (object != null && object == object && status == "enabled") {
+            res.json(object);
+        } else {
+            var body = ({});
+            var headers = {APP_ID: APP_ID};
+            var url = URL + '/home/slider/';
```
Comment (excellencetechnologies) [inline routes/home.js]: we need to resize slider images as well

File: routes/product.js [hunk 8]
```diff
@@ -64,7 +66,7 @@ router.post('/review', function (req, res) {
                     } else {
                         client.hmset('product_' + sku, {
```
Comment (excellencetechnologies) [inline routes/product.js]: we need to apply resize on product images as well

File: service/request.js [hunk 9]
```diff
@@ -1,7 +1,12 @@
 var request = require('request');
 var sharp = require('sharp');
 var http = require('http');
+var imagemin = require('imagemin');
+var imageminMozjpeg = require('imagemin-mozjpeg');
+var imageminPngquant = require('imagemin-pngquant');
 var fs = require('fs');
+var URL_ = require('url');
+var mkdirp = require('mkdirp');
```
Comment (excellencetechnologies) [inline service/request.js]: this is again to added to package.json!!!!

File: service/request.js [hunk 10]
```diff
@@ -16,10 +21,91 @@ exports.request = function (body, headers, url, callback) {
...
+                http.get(url, function (response) {
+                    response.pipe(file);
+                    response.on('end', function () {
+                        sharp('public/' + image_name)
+                                .resize(300, 200)
+                                .toFile('public/' + image_stored_url, function (err) {
+                                    if (err) {
+                                        callback(500, err);
+                                    } else if (err === null) {
+                                        callback(200, "done", 'http://144.76.34.244:8080/shekhar_works/Eexpress_magento/public/'+image_stored_url);
```
Comment (excellencetechnologies) [inline service/request.js]: put this in config

File: routes/category.js [hunk 11]
```diff
@@ -12,64 +10,74 @@ var redis = require("redis"),
...
+                                item.data.media_images[0] = image_url;
```
Comment (excellencetechnologies) [inline routes/category.js]: check error status here. if error then don't see new image_name because it will be null

File: service/request.js [hunk 12]
```diff
@@ -16,10 +21,91 @@ exports.request = function (body, headers, url, callback) {
...
+            } else {
+                callback(200, "done", 'http://144.76.34.244:8080/shekhar_works/Eexpress_magento/public/'+image_stored_url);
+            }
+        })
+    } else {
+        callback(500, " APP_ID or url cannot be empty");
+    }
+};
+
+exports.minify = function (url, APP_ID, callback) {
```
Comment (excellencetechnologies) [inline service/request.js]: this is not called anywhere right now?

File: service/request.js [hunk 13]
```diff
@@ -16,10 +21,91 @@ exports.request = function (body, headers, url, callback) {
...
+};
+
+exports.resize = function (url, APP_ID, callback) {
+
+    if (url.length > 0 && APP_ID.length > 0) {
+        var image_url = URL_.parse(url).path;
+        var app_id = APP_ID.replace(/[^a-zA-Z0-9 ]/g, "");
+        var image_stored_url = app_id + image_url;
+        var url_last_index_length = url.lastIndexOf('/');
+        var image_name = url.substring(url_last_index_length + 1);
```
Comment (excellencetechnologies) [inline service/request.js]: we are always resize image, if image is resize already it should not be done again

File: routes/category.js (hunk 3 above)
Comment (excellencetechnologies) [inline routes/category.js]: done rename changed to categoryData

File: routes/home.js (hunk 6 above)
Comment (excellencetechnologies) [inline routes/home.js]: use async here. now

File: service/request.js (hunk 9 above)
Comment (excellencetechnologies) [inline service/request.js]: added in package.json

File: service/request.js (hunk 10 above)
Comment (excellencetechnologies) [inline service/request.js]: put image_url in config file , done

File: service/request.js (hunk 13 above)
Comment (excellencetechnologies) [inline service/request.js]: already written code or this .

**Pull request merged**